app/scripts/lib/crypto/relier-keys.js

```diff
@@ -0,0 +1,76 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+/**
+ * Derive relier-specific encryption keys from account master keys.
+ */
+
+define(function (require, exports, module) {
+  'use strict';
+
+  const requireOnDemand = require('lib/require-on-demand');
+  const p = require('lib/promise');
+
+  /**
+   *
+   * Given the account master key kB, generate the matching relier-specific derived
+   * scoped key.
+   *
+   * @param {Object} keys - Account keys, used to derive scoped keys
+   * @param {Object} keyData - OAuth client data that is required to derive keys
+   * @returns {Promise} A promise that will resolve with an object having a scoped key.
+   *   The key is represented as a JWK object.
+   */
+  function _deriveRelierKeys(keys = {}, keyData = {}) {
+    return requireOnDemand('fxaCryptoDeriver').then((fxaCryptoDeriver) => {
+      const scopedKeys = new fxaCryptoDeriver.ScopedKeys();
+
+      return scopedKeys.deriveScopedKeys({
+        identifier: keyData.identifier,
+        inputKey: keys.kB,
+        keyMaterial: keyData.keyMaterial,
+        timestamp: keyData.timestamp
+      });
+    });
+  }
+
+  /**
+   * Derive scoped keys and create an encrypted bundle for key transport
+   *
+   * @param {Object} keys - Account keys, used to derive scoped keys
+   * @param {Object} clientKeyData - OAuth client data that is required to derive keys
+   * @param {Object} keysJwk - Public key used for scoped key encryption
+   * @returns {Promise} A promise that will resolve into an encrypted bundle of scoped keys
+   */
+  function createEncryptedBundle(keys, clientKeyData, keysJwk) {
+    const relierKeys = [];
+    const clientKeyDataScopes = Object.keys(clientKeyData);
+
+    clientKeyDataScopes.forEach((key) => {
+      relierKeys.push(_deriveRelierKeys(keys, clientKeyData[key]));
+    });
+
+    return p.all(relierKeys)
+      .then((derivedKeys) => {
+        const scopedKeys = {};
+
+        derivedKeys.forEach((item) => {
+          const scopeName = Object.keys(item)[0];
```

Comment:
Is there only ever a single key/value pair in `item`? If not, assuming `scopeName` is always the first key in the object is dangerous. Firefox keeps ordering, I read Chrome usually does but makes no guarantees.

Reply:
> Is there only ever a single key/value pair in item?

Yea there will always be one.